From 3fd3ed156ec9554a816ebf5bf15159ecffc2823c Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Sun, 15 Jun 2025 13:13:26 +0300
Subject: [PATCH] platform/mellanox: nvsw-host-spc5: Add system control and
 monitoring driver for Nvidia SPC5 switches

The driver allows system control and monitoring of Nvidia SPC5 based
switches through CPLD/FPGA devices from the host CPU.

This control includes:
- Handling of hotplug events.
- Various resets handling through sysfs attributes.
- Reset cause reports through sysfs attributes.
- Mux attributes for hw components selection.
- General purpose registers attributes.
- LED control.
- FAN control.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/platform/mellanox/Kconfig          |  10 +
 drivers/platform/mellanox/Makefile         |   1 +
 drivers/platform/mellanox/nvsw-host-spc5.c | 952 +++++++++++++++++++++
 3 files changed, 963 insertions(+)
 create mode 100644 drivers/platform/mellanox/nvsw-host-spc5.c

diff --git a/drivers/platform/mellanox/Kconfig b/drivers/platform/mellanox/Kconfig
index 985cd2f..8b8e925 100644
--- a/drivers/platform/mellanox/Kconfig
+++ b/drivers/platform/mellanox/Kconfig
@@ -123,6 +123,16 @@ config NVSW_HOST_NVL6
 	  This driver can also be built as a module. If so the module
 	  will be called nvsw-host-nvl6.
 
+config NVSW_HOST_SPC5
+	tristate "Nvidia SPC5 host CPLD/FPGA Hardware Control and Monitoring"
+	depends on HWMON && I2C
+	select REGMAP_I2C
+	help
+	  Say Y here to include support for the FPGA/CPLD logic by host CPU
+	  equipped on Nvidia SPC5 ethernet switches.
+	  This driver can also be built as a module. If so the module
+	  will be called nvsw-host-spc5.
+
 config NVSW_HOST_SPC6
 	tristate "Nvidia SPC6 host CPLD/FPGA Hardware Control and Monitoring"
 	depends on HWMON && I2C
diff --git a/drivers/platform/mellanox/Makefile b/drivers/platform/mellanox/Makefile
index 17cc4fd..5f8e9a5 100644
--- a/drivers/platform/mellanox/Makefile
+++ b/drivers/platform/mellanox/Makefile
@@ -12,6 +12,7 @@ obj-$(CONFIG_MLXREG_HOTPLUG)	+= mlxreg-hotplug.o
 obj-$(CONFIG_MLXREG_IO) += mlxreg-io.o
 obj-$(CONFIG_MLXREG_LC) += mlxreg-lc.o
 obj-$(CONFIG_NVSW_HOST_NVL6) += nvsw-host-nvl6.o nvsw-core.o
+obj-$(CONFIG_NVSW_HOST_SPC5) += nvsw-host-spc5.o nvsw-core.o
 obj-$(CONFIG_NVSW_HOST_SPC6) += nvsw-host-spc6.o nvsw-core.o
 obj-$(CONFIG_NVSW_SN2201) += nvsw-sn2201.o
 obj-$(CONFIG_MLXREG_DPU)	+= mlxreg-dpu.o
diff --git a/drivers/platform/mellanox/nvsw-host-spc5.c b/drivers/platform/mellanox/nvsw-host-spc5.c
new file mode 100644
index 0000000..54ba33c
--- /dev/null
+++ b/drivers/platform/mellanox/nvsw-host-spc5.c
@@ -0,0 +1,952 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Nvidia SPC5 host platform driver
+ *
+ * Copyright (C) 2025 Nvidia Technologies Ltd.
+ */
+
+#include <linux/device.h>
+#include <linux/dmi.h>
+#include <linux/i2c.h>
+#include <linux/i2c-mux.h>
+#include <linux/io.h>
+#include <linux/module.h>
+#include <linux/pci.h>
+#include <linux/platform_device.h>
+#include <linux/platform_data/i2c-mux-reg.h>
+#include <linux/platform_data/mlxreg.h>
+#include <linux/regmap.h>
+
+#include "nvsw.h"
+
+#define NVSW_HOST_DEVICE_NAME	"mlxplat"
+
+/* LPC bus IO offsets */
+#define NVSW_I2C_BASE_ADRR	0x2000
+#define NVSW_REG_BASE_ADRR	0x2500
+#define NVSW_LPC_IO_RANGE	0x100
+#define NVSW_LPC_PIO_OFFSET	0x10000UL
+#define NVSW_REG_MUX1		(NVSW_REG_MUX1_OFFSET | NVSW_LPC_PIO_OFFSET)
+#define NVSW_REG_MUX2		(NVSW_REG_MUX0_OFFSET | NVSW_LPC_PIO_OFFSET)
+
+/* Default I2C parent bus number */
+#define NVSW_CPLD_PHYS_ADAPTER_DEF_NR	5
+
+/* Start channel numbers */
+#define NVSW_PARENT_CH		NVSW_CPLD_PHYS_ADAPTER_DEF_NR
+#define NVSW_MUX1_CH		(NVSW_CPLD_PHYS_ADAPTER_DEF_NR + 1)
+#define NVSW_MUX2_CH		64
+
+
+/* Regions for LPC I2C controller and LPC base register space */
+static const struct resource nvsw_host_io_resources[] = {
+	[0] = DEFINE_RES_NAMED(NVSW_I2C_BASE_ADRR, NVSW_LPC_IO_RANGE,
+			       "nvsw_cpld_lpc_i2c_ctrl", IORESOURCE_IO),
+	[1] = DEFINE_RES_NAMED(NVSW_REG_BASE_ADRR, NVSW_LPC_IO_RANGE,
+			       "nvsw_cpld_lpc_regs", IORESOURCE_IO),
+};
+
+/* SPC5 platform mgmt board mux channels */
+static const int nvsw_host_spc5_mgmt_channels[] = {
+	1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17,
+	18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+	33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
+	48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58
+};
+
+/* SPC5 platform ComEx board mux channels */
+static const int nvsw_host_spc5_comex_channels[] = {
+	1, 2, 3, 4, 5, 6, 7, 8
+};
+
+/* SPC5 platform mux data */
+static struct i2c_mux_reg_platform_data nvsw_host_spc5_mux_data[] = {
+	{
+		.parent = NVSW_PARENT_CH,
+		.base_nr = NVSW_MUX1_CH,
+		.write_only = 1,
+		.reg = (void __iomem *)NVSW_REG_MUX1,
+		.reg_size = 1,
+		.idle_in_use = 1,
+		.values = nvsw_host_spc5_mgmt_channels,
+		.n_values = ARRAY_SIZE(nvsw_host_spc5_mgmt_channels),
+	},
+		{
+		.parent = NVSW_PARENT_CH,
+		.base_nr = NVSW_MUX2_CH,
+		.write_only = 1,
+		.reg = (void __iomem *)NVSW_REG_MUX2,
+		.reg_size = 1,
+		.idle_in_use = 1,
+		.values = nvsw_host_spc5_comex_channels,
+		.n_values = ARRAY_SIZE(nvsw_host_spc5_comex_channels),
+	},
+};
+
+/* SPC5 platform i2c frequency data */
+static struct mlxreg_core_data nvsw_mlxcpld_i2c_items_data[] = {
+	{
+		.reg = NVSW_REG_PSU_I2C_CAP_OFFSET,
+		.mask = NVSW_I2C_CAP_MASK,
+		.bit = NVSW_I2C_CAP_BIT,
+	},
+};
+
+static struct mlxreg_core_item nvsw_mlxcpld_i2c_items[] = {
+	{
+		.data = nvsw_mlxcpld_i2c_items_data,
+	},
+};
+
+/* SPC5 platform i2c data */
+static struct mlxreg_core_hotplug_platform_data nvsw_mlxcpld_i2c_data = {
+	.items = nvsw_mlxcpld_i2c_items,
+	.cell = NVSW_REG_AGGR_OFFSET,
+	.mask = NVSW_AGGR_MASK_COMEX,
+	.cell_low = NVSW_REG_AGGRCO_OFFSET,
+	.mask_low = NVSW_LOW_AGGR_MASK_I2C,
+};
+
+static struct platform_device *nvsw_host_dev;
+static struct platform_device *nvsw_i2c_dev;
+static struct i2c_mux_reg_platform_data *nvsw_host_mux_data[NVSW_MUX_MAX];
+static struct mlxreg_core_platform_data *nvsw_led_data;
+static struct mlxreg_core_platform_data *nvsw_regs_io_data;
+static struct mlxreg_core_hotplug_platform_data *nvsw_i2c_data;
+static struct mlxreg_core_platform_data *nvsw_wd_data[NVSW_WD_MAX];
+static int mux_num;
+static enum nvsw_core_hid_type nvsw_host_hid;
+
+/* Platform register access for SPC5 systems families data */
+static struct mlxreg_core_data nvsw_host_spc5_regs_io_data[] = {
+	{
+		.label = "cpld1_version",
+		.reg = NVSW_REG_CPLD1_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld2_version",
+		.reg = NVSW_REG_CPLD2_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld3_version",
+		.reg = NVSW_REG_CPLD3_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld4_version",
+		.reg = NVSW_REG_CPLD4_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld5_version",
+		.reg = NVSW_REG_CPLD5_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld6_version",
+		.reg = NVSW_REG_CPLD6_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld7_version",
+		.reg = NVSW_REG_CPLD7_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld8_version",
+		.reg = NVSW_REG_CPLD8_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld9_version",
+		.reg = NVSW_REG_CPLD9_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld10_version",
+		.reg = NVSW_REG_CPLD10_VER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld1_pn",
+		.reg = NVSW_REG_CPLD1_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld2_pn",
+		.reg = NVSW_REG_CPLD2_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld3_pn",
+		.reg = NVSW_REG_CPLD3_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld4_pn",
+		.reg = NVSW_REG_CPLD4_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld5_pn",
+		.reg = NVSW_REG_CPLD5_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld6_pn",
+		.reg = NVSW_REG_CPLD6_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld7_pn",
+		.reg = NVSW_REG_CPLD7_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld8_pn",
+		.reg = NVSW_REG_CPLD8_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld9_pn",
+		.reg = NVSW_REG_CPLD9_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld10_pn",
+		.reg = NVSW_REG_CPLD10_PN_OFFSET,
+		.bit = GENMASK(15, 0),
+		.mode = 0444,
+		.regnum = 2,
+	},
+	{
+		.label = "cpld1_version_min",
+		.reg = NVSW_REG_CPLD1_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld2_version_min",
+		.reg = NVSW_REG_CPLD2_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld3_version_min",
+		.reg = NVSW_REG_CPLD3_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld4_version_min",
+		.reg = NVSW_REG_CPLD4_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld5_version_min",
+		.reg = NVSW_REG_CPLD5_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld6_version_min",
+		.reg = NVSW_REG_CPLD6_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld7_version_min",
+		.reg = NVSW_REG_CPLD7_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld8_version_min",
+		.reg = NVSW_REG_CPLD8_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld9_version_min",
+		.reg = NVSW_REG_CPLD9_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "cpld10_version_min",
+		.reg = NVSW_REG_CPLD10_MVER_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "pwr_converter_prog_en",
+		.reg = NVSW_REG_GP7_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0644,
+	},
+	{
+		.label = "graceful_pwr_off",
+		.reg = NVSW_REG_GP7_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_mctp_ready",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_shutdown_req",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "vpd_wp",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+		.secured = 1,
+	},
+	{
+		.label = "pcie_asic_reset_dis",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0644,
+	},
+	{
+		.label = "shutdown_unlock",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "cpu_power_off_ready",
+		.reg = NVSW_REG_GP0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0644,
+	},
+	{
+		.label = "pwr_cycle",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0244,
+	},
+	{
+		.label = "pwr_down",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0200,
+	},
+	{
+		.label = "aux_pwr_cycle",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0200,
+	},
+	{
+		.label = "bmc_to_cpu_ctrl",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "uart_sel",
+		.reg = NVSW_REG_GP1_OFFSET,
+		.mask = NVSW_UART_SEL_MASK,
+		.bit = 7,
+		.mode = 0644,
+	},
+	{
+		.label = "leakage1",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage2",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage3",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage4",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "leakage5",
+		.reg = NVSW_REG_LEAK_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0444,
+	},
+	{
+		.label = "asic_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+	{
+		.label = "asic1_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0644,
+	},
+	{
+		.label = "asic2_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0644,
+	},
+	{
+		.label = "asic3_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0644,
+	},
+	{
+		.label = "asic4_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0644,
+	},
+	{
+		.label = "sgmii_phy_reset",
+		.reg = NVSW_REG_RESET_GP2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0644,
+	},
+	{
+		.label = "reset_long_pb",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_short_pb",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_aux_pwr_or_reload",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_swb_dc_dc_pwr_fail",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_platform",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(4),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_swb_wd",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_asic_thermal",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_comex_pwr_fail",
+		.reg = NVSW_REG_RESET_CAUSE1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_soc",
+		.reg = NVSW_REG_RESET_CAUSE1_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_system",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_sw_pwr_off",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_comex_thermal",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_pwr_converter_fail",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_main_51v",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.mode = 0444,
+	},
+	{
+		.label = "reset_mgmt_pwr",
+		.reg = NVSW_REG_RESET_CAUSE2_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "port80",
+		.reg = NVSW_REG_GP1_RO_OFFSET,
+		.bit = GENMASK(7, 0),
+		.mode = 0444,
+	},
+	{
+		.label = "bios_active_image",
+		.reg = NVSW_REG_GPCOM0_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(5),
+		.mode = 0444,
+	},
+	{
+		.label = "jtag_cap",
+		.reg = NVSW_REG_FU_CAP_OFFSET,
+		.mask = NVSW_FU_CAP_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "jtag_enable",
+		.reg = NVSW_REG_FIELD_UPGRADE,
+		.mask = GENMASK(1, 0),
+		.bit = 1,
+		.mode = 0644,
+	},
+	{
+		.label = "asic_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic2_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC2_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic3_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC3_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic4_health",
+		.reg = NVSW_REG_ASIC1_HEALTH_OFFSET,
+		.mask = NVSW_ASIC4_MASK,
+		.bit = 1,
+		.mode = 0444,
+	},
+	{
+		.label = "asic1_pg_fail",
+		.reg = NVSW_REG_GP4_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "asic2_pg_fail",
+		.reg = NVSW_REG_GP4_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "asic3_pg_fail",
+		.reg = NVSW_REG_GP5_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(7),
+		.mode = 0444,
+	},
+	{
+		.label = "asic4_pg_fail",
+		.reg = NVSW_REG_GP5_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0444,
+	},
+	{
+		.label = "spi_chnl_select",
+		.reg = NVSW_REG_SPI_CHNL_SELECT,
+		.mask = GENMASK(7, 0),
+		.bit = 1,
+		.mode = 0644,
+	},
+	{
+		.label = "pdb1_pwr_status",
+		.reg = NVSW_REG_GP0_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(0),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb2_pwr_status",
+		.reg = NVSW_REG_GP0_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(1),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb3_pwr_status",
+		.reg = NVSW_REG_GP0_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(2),
+		.mode = 0644,
+	},
+	{
+		.label = "pdb4_pwr_status",
+		.reg = NVSW_REG_GP0_RO_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(3),
+		.mode = 0644,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_spc5_regs_io = {
+	.data = nvsw_host_spc5_regs_io_data,
+	.counter = ARRAY_SIZE(nvsw_host_spc5_regs_io_data),
+};
+
+/* SPC5 platform led data */
+static struct mlxreg_core_data nvsw_host_spc5_led_data[] = {
+	{
+		.label = "status:green",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+	{
+		.label = "status:amber",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+	{
+		.label = "power:green",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_HI_NIBBLE_MASK,
+	},
+	{
+		.label = "power:amber",
+		.reg = NVSW_REG_LED1_OFFSET,
+		.mask = NVSW_LED_HI_NIBBLE_MASK,
+	},
+	{
+		.label = "uid:blue",
+		.reg = NVSW_REG_LED5_OFFSET,
+		.mask = NVSW_LED_LO_NIBBLE_MASK,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_spc5_led = {
+	.data = nvsw_host_spc5_led_data,
+	.counter = ARRAY_SIZE(nvsw_host_spc5_led_data),
+};
+
+/* Watchdog type3 platform data */
+static struct mlxreg_core_data nvsw_host_wd_main_regs_type3[] = {
+	{
+		.label = "action",
+		.reg = NVSW_REG_WD2_ACT_OFFSET,
+		.mask = NVSW_WD_RESET_ACT_MASK,
+		.bit = 0,
+	},
+	{
+		.label = "timeout",
+		.reg = NVSW_REG_WD2_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+		.health_cntr = NVSW_WD3_DFLT_TIMEOUT,
+	},
+	{
+		.label = "timeleft",
+		.reg = NVSW_REG_WD2_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+	},
+	{
+		.label = "ping",
+		.reg = NVSW_REG_WD2_ACT_OFFSET,
+		.mask = NVSW_WD_RESET_ACT_MASK,
+		.bit = 0,
+	},
+	{
+		.label = "reset",
+		.reg = NVSW_REG_RESET_CAUSE_OFFSET,
+		.mask = GENMASK(7, 0) & ~BIT(6),
+		.bit = 6,
+	},
+};
+
+static struct mlxreg_core_data nvsw_host_wd_aux_regs_type3[] = {
+	{
+		.label = "action",
+		.reg = NVSW_REG_WD3_ACT_OFFSET,
+		.mask = NVSW_WD_FAN_ACT_MASK,
+		.bit = 4,
+	},
+	{
+		.label = "timeout",
+		.reg = NVSW_REG_WD3_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+		.health_cntr = NVSW_WD3_DFLT_TIMEOUT,
+	},
+	{
+		.label = "timeleft",
+		.reg = NVSW_REG_WD3_TMR_OFFSET,
+		.mask = NVSW_WD_TYPE2_TO_MASK,
+	},
+	{
+		.label = "ping",
+		.reg = NVSW_REG_WD3_ACT_OFFSET,
+		.mask = NVSW_WD_FAN_ACT_MASK,
+		.bit = 4,
+	},
+};
+
+static struct mlxreg_core_platform_data nvsw_host_wd_set_type3[] = {
+	{
+		.data = nvsw_host_wd_main_regs_type3,
+		.counter = ARRAY_SIZE(nvsw_host_wd_main_regs_type3),
+		.version = MLX_WDT_TYPE3,
+		.identity = "mlx-wdt-main",
+	},
+	{
+		.data = nvsw_host_wd_aux_regs_type3,
+		.counter = ARRAY_SIZE(nvsw_host_wd_aux_regs_type3),
+		.version = MLX_WDT_TYPE3,
+		.identity = "mlx-wdt-aux",
+	},
+};
+
+/* IO port mapping callback. */
+static void __iomem *nvsw_host_spc5_port_map(struct nvsw_core *nvsw_core)
+{
+	return devm_ioport_map(nvsw_core->dev, nvsw_host_io_resources[1].start, 1);
+}
+
+/* Mux init/exit callbacks. */
+static int nvsw_host_spc5_mux_topology_init(struct nvsw_core *nvsw_core)
+{
+	int i, err;
+
+	if (!nvsw_i2c_data)
+		return 0;
+
+	/* Create i2c infrastructure. */
+	nvsw_i2c_data->regmap = nvsw_core->regmap;
+	nvsw_i2c_data->handle = nvsw_core;
+
+	nvsw_i2c_dev = platform_device_register_resndata(nvsw_core->dev, "i2c_mlxcpld",
+							   NVSW_CPLD_PHYS_ADAPTER_DEF_NR, NULL, 0,
+							   nvsw_i2c_data, sizeof(*nvsw_i2c_data));
+	if (IS_ERR(nvsw_i2c_dev)) {
+		err = PTR_ERR(nvsw_i2c_dev);
+		nvsw_i2c_dev = NULL;
+		return err;
+	}
+
+	/* Create mux infrastructure. */
+	for (i = 0; i < nvsw_core->mux_num; i++) {
+		nvsw_core->mux[i] =
+			platform_device_register_resndata(nvsw_core->dev, "i2c-mux-reg", i, NULL,
+							  0, nvsw_host_mux_data[i],
+							  sizeof(*nvsw_host_mux_data[i]));
+		if (IS_ERR(nvsw_core->mux[i])) {
+			dev_err(nvsw_core->dev, "Failed to create mux infra\n");
+			err = PTR_ERR(nvsw_core->mux[i]);
+			nvsw_core->mux[i] = NULL;
+			goto fail_platform_mux_register;
+		}
+	}
+
+	return 0;
+fail_platform_mux_register:
+	while (--i >= 0) {
+		platform_device_unregister(nvsw_core->mux[i]);
+		nvsw_core->mux[i] = NULL;
+	}
+	platform_device_unregister(nvsw_i2c_dev);
+	nvsw_i2c_dev = NULL;
+	return err;
+}
+
+static void nvsw_host_spc5_mux_topology_exit(struct nvsw_core *nvsw_core)
+{
+	int i;
+
+	for (i = 0; i < nvsw_core->mux_num; i++) {
+		if (nvsw_core->mux[i]) {
+			platform_device_unregister(nvsw_core->mux[i]);
+			nvsw_core->mux[i] = NULL;
+		}
+	}
+
+	if (nvsw_i2c_dev) {
+		platform_device_unregister(nvsw_i2c_dev);
+		nvsw_i2c_dev = NULL;
+	}
+}
+
+static int __init nvsw_host_dmi_spc5_switch_matched(const struct dmi_system_id *dmi)
+{
+	int i;
+
+	/* Set system configuration. */
+	nvsw_host_hid = HID181;
+	mux_num = ARRAY_SIZE(nvsw_host_spc5_mux_data);
+	for (i = 0; i < mux_num; i++)
+		nvsw_host_mux_data[i] = &nvsw_host_spc5_mux_data[i];
+	nvsw_led_data = &nvsw_host_spc5_led;
+	nvsw_regs_io_data = &nvsw_host_spc5_regs_io;
+	for (i = 0; i < ARRAY_SIZE(nvsw_host_wd_set_type3); i++)
+		nvsw_wd_data[i] = &nvsw_host_wd_set_type3[i];
+	nvsw_i2c_data = &nvsw_mlxcpld_i2c_data;
+
+	return 1;
+}
+
+static const struct dmi_system_id nvsw_host_dmi_table[] __initconst = {
+	{
+		.callback = nvsw_host_dmi_spc5_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0024"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI181"),
+		},
+	},
+	{
+		.callback = nvsw_host_dmi_spc5_switch_matched,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_NAME, "VMOD0024"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_SKU, "HI182"),
+		},
+	},
+	{ }
+};
+
+MODULE_DEVICE_TABLE(dmi, nvsw_host_dmi_table);
+
+static int nvsw_host_probe(struct platform_device *pdev)
+{
+	struct nvsw_core *nvsw_core;
+	int i;
+
+	nvsw_core = devm_kzalloc(&nvsw_host_dev->dev, sizeof(*nvsw_core), GFP_KERNEL);
+	if (!nvsw_core)
+		return -ENOMEM;
+
+	/* Set system configuration. */
+	nvsw_core->dev = &nvsw_host_dev->dev;
+	nvsw_core->hid = nvsw_host_hid;
+	nvsw_core->regmap_type = REGMAP_IO;
+	nvsw_core->mux_num = mux_num;
+	for (i = 0; i < ARRAY_SIZE(nvsw_wd_data); i++)
+		nvsw_core->wd_data[i] = nvsw_wd_data[i];
+	nvsw_core->regio_data = nvsw_regs_io_data;
+	nvsw_core->led_data = nvsw_led_data;
+	nvsw_core->port_map = nvsw_host_spc5_port_map;
+	nvsw_core->mux_init = nvsw_host_spc5_mux_topology_init;
+	nvsw_core->mux_exit = nvsw_host_spc5_mux_topology_exit;
+	platform_set_drvdata(nvsw_host_dev, nvsw_core);
+
+	return nvsw_core_init(nvsw_core);
+}
+
+static int nvsw_host_remove(struct platform_device *pdev)
+{
+	struct nvsw_core *nvsw_core = platform_get_drvdata(nvsw_host_dev);
+
+	nvsw_core_exit(nvsw_core);
+	return 0;
+}
+
+static struct platform_driver nvsw_host_driver = {
+	.driver		= {
+		.name	= NVSW_HOST_DEVICE_NAME,
+		.probe_type = PROBE_FORCE_SYNCHRONOUS,
+	},
+	.probe		= nvsw_host_probe,
+	.remove		= nvsw_host_remove,
+};
+
+static int __init nvsw_host_init(void)
+{
+	int err;
+
+	if (!dmi_check_system(nvsw_host_dmi_table))
+		return -ENODEV;
+
+	nvsw_host_dev = platform_device_register_simple(NVSW_HOST_DEVICE_NAME, -1,
+							nvsw_host_io_resources,
+							ARRAY_SIZE(nvsw_host_io_resources));
+	if (IS_ERR(nvsw_host_dev))
+		return PTR_ERR(nvsw_host_dev);
+
+	err = platform_driver_register(&nvsw_host_driver);
+	if (err)
+		platform_device_unregister(nvsw_host_dev);
+
+	return err;
+}
+module_init(nvsw_host_init);
+
+static void __exit nvsw_host_exit(void)
+{
+	platform_device_unregister(nvsw_host_dev);
+	platform_driver_unregister(&nvsw_host_driver);
+}
+module_exit(nvsw_host_exit);
+
+MODULE_AUTHOR("Felix Radensky <fradensky@nvidia.com>");
+MODULE_DESCRIPTION("Nvidia platform driver");
+MODULE_LICENSE("Dual BSD/GPL");
-- 
2.34.1

